sycl/source/detail/device_image_impl.hpp

```diff
@@ -0,0 +1,136 @@
+//==------- device_image_impl.hpp - SYCL device_image_impl -----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/context.hpp>
+#include <CL/sycl/detail/common.hpp>
+#include <CL/sycl/detail/pi.h>
+#include <CL/sycl/device.hpp>
+#include <CL/sycl/kernel_bundle.hpp>
+#include <detail/device_impl.hpp>
+#include <detail/kernel_id_impl.hpp>
+#include <detail/program_manager/program_manager.hpp>
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+#include <memory>
+#include <vector>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+// The class is impl counterpart for sycl::image_impl
+// It can represent a program in different states, kernel_id's it has and state
+// of specialization constants for it
+class device_image_impl {
+public:
+  device_image_impl(RTDeviceBinaryImage *BinImage, context Context,
+                    std::vector<device> Devices, bundle_state State)
+      : MBinImage(BinImage), MContext(std::move(Context)),
+        MDevices(std::move(Devices)), MState(State) {
+
+    // Collect kernel names for the image
+    pi_device_binary DevBin =
+        const_cast<pi_device_binary>(&BinImage->getRawData());
+    for (_pi_offload_entry EntriesIt = DevBin->EntriesBegin;
+         EntriesIt != DevBin->EntriesEnd; ++EntriesIt) {
+
+      std::shared_ptr<detail::kernel_id_impl> KernleIDImpl =
+          std::make_shared<detail::kernel_id_impl>(EntriesIt->name);
+      MKernelIDs.emplace_back(
+          detail::createSyclObjFromImpl<sycl::kernel_id>(KernleIDImpl));
+    }
+  }
+
+  bool has_kernel(const kernel_id &KernelIDCand) const noexcept {
+    return std::any_of(MKernelIDs.begin(), MKernelIDs.end(),
+                       [&KernelIDCand](const kernel_id &KernelID) {
+                         return strcmp(KernelID.get_name(),
+                                       KernelIDCand.get_name()) == 0;
+                       });
+  }
+
+  bool has_kernel(const kernel_id &KernelIDCand,
+                  const device &DeviceCand) const noexcept {
+    for (const device &Device : MDevices)
+      if (Device == DeviceCand)
+        return has_kernel(KernelIDCand);
+
+    return false;
+  }
+
+  const std::vector<kernel_id> &get_kernel_ids() const noexcept {
+    return MKernelIDs;
+  }
+
+  bool has_specialization_constants() const noexcept {
+    return !MSpecConstsBlob.empty();
+  }
+
+  bool all_specialization_constant_native() const noexcept {
+    assert(false && "Not implemented");
+    return false;
+  }
+
+  // The struct maps specialization ID to offset in the binary blob where value
+  // for this spec const should be.
+  struct SpecConstIDOffset {
```

Comment:
I see spec constants-related stuff in this patch, on the other hand @AlexeySachkov is still working on SYCL2020 spec constants support design. Can you guys comment on your approach here - why is this being implemented before the design finishes?

Reply:
The mentions of spec constants which are added in the patch do not enable any support. They just add a couple of placeholders which, I believe, will be needed anyway as part of kernel_bundle support.